### PR TITLE
Factory: adds new query to search for a liquidity pool address by a tuple of Pair. Tests the new functionality.

### PR DESCRIPTION
**Pull request by gangov**

### CODE REVIEW

File: contracts/factory/src/contract.rs [hunk 1]
```diff
@@ -125,6 +131,34 @@ impl FactoryTrait for Factory {
...
+            token_b: tuple_pair.1.clone(),
+        });
+
+        if let Some(addr) = pair_result {
+            return Ok(addr);
+        }
+
+        // this would've been the use case if we can swap memory without std
+        // std::mem::swap(&mut tuple_pair.0, &mut tuple_pair.1);
+
```
Comment (ueco-jb) [inline contracts/factory/src/contract.rs:149]:
```suggestion
```

File: contracts/factory/src/storage.rs [hunk 2]
```diff
@@ -68,6 +76,16 @@ pub fn save_lp_vec(env: &Env, lp_info: Vec<Address>) {
     env.storage().instance().set(&DataKey::LpVec, &lp_info);
 }
 
+pub fn save_lp_vec_with_tuple_as_key(env: &Env, tuple_pair: &TokenInitInfo, lp_address: &Address) {
```
Comment (ueco-jb) [inline contracts/factory/src/storage.rs:79]: Reference is cheap but it's still an overkill to pass the whole structure, it's enough to pass `(Address, Address)` or something similar - then you won't have to import the whole structure.

File: contracts/factory/src/tests/queries.rs [hunk 3]
```diff
@@ -221,4 +222,183 @@ fn test_deploy_multiple_liquidity_pools() {
...
+    let first_lp_config: Config = env.invoke_contract(
+        &lp_contract_addr,
+        &Symbol::new(&env, "query_config"),
+        Vec::new(&env),
+    );
+
+    assert_eq!(
+        first_lp_init_info.max_allowed_spread_bps,
+        first_lp_config.max_allowed_spread_bps
+    );
```
Comment (ueco-jb) [inline contracts/factory/src/tests/queries.rs:354]: There's no point in testing those addresses in this particular case. If you keep test cases to essentials then it's easier to read and to not overlook potential errors. 

```suggestion
```

File: contracts/factory/src/tests/queries.rs [hunk 4]
```diff
@@ -221,4 +222,183 @@ fn test_deploy_multiple_liquidity_pools() {
...
+    assert_eq!(token4, second_result.pool_response.asset_b.address);
+    assert_eq!(
+        second_share_token_addr,
+        second_result.pool_response.asset_lp_share.address
+    );
+    assert_eq!(second_lp_contract_addr, second_result.pool_address);
+
+    let third_lp_contract_addr = factory.query_pools().get(2).unwrap();
+
+    let first_lp_address_by_tuple = factory.query_for_pool_by_pair_tuple(&(token1, token2));
```
Comment (ueco-jb) [inline contracts/factory/src/tests/queries.rs:392]: You tested only a case with proper order of addresses. 

```suggestion
    let first_lp_address_by_tuple = factory.query_for_pool_by_pair_tuple(&(token2, token1));
```

**Review by ueco-jb — COMMENTED** [review]
Good code and good test. I added some comments to remove some stuff, especially trim that test case.

File: contracts/factory/src/tests/queries.rs [hunk 5]
```diff
@@ -221,4 +222,164 @@ fn test_deploy_multiple_liquidity_pools() {
...
+    let third_lp_address_by_tuple = factory.query_for_pool_by_pair_tuple(&(token5, token6));
+
+    assert_eq!(first_lp_address_by_tuple, lp_contract_addr);
+    assert_eq!(second_lp_address_by_tuple, second_lp_contract_addr);
+    assert_eq!(third_lp_address_by_tuple, third_lp_contract_addr);
+
+    assert_eq!(
+        factory.try_query_for_pool_by_pair_tuple(&(Address::random(&env), Address::random(&env))),
+        Err(Ok(ContractError::LiquidityPoolPairNotFound))
+    );
```
Comment (ueco-jb) [inline contracts/factory/src/tests/queries.rs:384]: For the future, checking random addresses probably doesn't make sense. What you should check is a pair of existing tokens, something like:
```suggestion
    assert_eq!(
        factory.try_query_for_pool_by_pair_tuple(&(token2, token3)),
        Err(Ok(ContractError::LiquidityPoolPairNotFound))
    );
```

**Review by ueco-jb — APPROVED** [review]
LGTM

File: contracts/factory/src/tests/queries.rs (hunk 5 above)
Comment (gangov) [reply contracts/factory/src/tests/queries.rs:384]: got you

**Pull request merged**